pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -231,21 +234,30 @@ public SendCallback getNextSendCallback() {
             public void sendComplete(Exception e) {
                 if (e != null) {
                     stats.incrementSendFailed();
+                    onSendAcknowledgement(interceptorMessage, null, e);
                     future.completeExceptionally(e);
                 } else {
-                    future.complete(message.getMessageId());
+                    onSendAcknowledgement(interceptorMessage, interceptorMessage.getMessageId(), null);
+                    future.complete(interceptorMessage.getMessageId());
                     stats.incrementNumAcksReceived(System.nanoTime() - createdAt);
                 }
+                interceptorMessage.getDataBuffer().release();
```

Comment:
It would be good to have try/finally block to ensure release is called when any exception is thrown